src/metabase/driver/h2.clj

```diff
@@ -282,6 +283,77 @@
+            (throw (ex-info (tru "Only datetime, timestamp, or date types allowed. Found {0}"
+                                 (pr-str disallowed-types))
+                            {:found disallowed-types
+                             :type  qp.error-type/invalid-query})))
+        x (hx/->timestamp x)
+        y (hx/->timestamp y)]
+    (case unit
+      :year
+      (let [positive-diff (fn [a b] ; precondition: a <= b
+                            (hx/-
+                             (hx/- (extract :year b) (extract :year a))
+                             ;; decrement if a is later than b in the year calendar
+                             (hx/cast
+                              :integer
+                              (hsql/call
+                               :or
+                               (hsql/call :> (extract :month a) (extract :month b))
+                               (hsql/call
+                                :and
+                                (hsql/call := (extract :month a) (extract :month b))
+                                (hsql/call :> (extract :day a) (extract :day b)))))))]
+        (hsql/call :case (hsql/call :<= x y) (positive-diff x y) :else (hx/* -1 (positive-diff y x))))
+
+      :quarter
+      (let [positive-diff
+            (fn [a b]
+              (hx/cast :integer (hx/floor (hx// (hx/- (hsql/call :datediff (hsql/raw "month") a b)
+                                                      (hx/cast :integer (hsql/call :> (extract :day a) (extract :day b))))
+                                                3))))]
+        (hsql/call :case (hsql/call :<= x y) (positive-diff x y) :else (hx/* -1 (positive-diff y x))))
+
+      :month
+      (let [positive-diff (fn [a b]
+                            (hx/-
+                             (hsql/call :datediff (hsql/raw "month") a b)
+                             (hx/cast :integer (hsql/call :> (extract :day a) (extract :day b)))))]
+        (hsql/call :case (hsql/call :<= x y) (positive-diff x y) :else (hx/* -1 (positive-diff y x))))
+
+      :week
+      (let [positive-diff (fn [a b]
+                            (hx/cast
+                             :integer
+                             (hx/floor (hx// (hsql/call :datediff (hsql/raw "day") a b) 7))))]
+        (hsql/call :case (hsql/call :<= x y) (positive-diff x y) :else (hx/* -1 (positive-diff y x))))
+
+      :day
+      (hsql/call :datediff (hsql/raw (name unit)) x y)
+
+      (:hour :minute :second)
+      (let [positive-diff (fn [a b]
+                            (hx/cast
+                             :integer
+                             (hx/floor
+                              (hx// (hx/cast :float (hsql/call :datediff (hsql/raw "millisecond") a b)) ; timestampdiff returns integer, so cast to float
+                                    (case unit :hour 3600000 :minute 60000 :second 1000)))))]
+        (hsql/call :case (hsql/call :<= x y) (positive-diff x y) :else (hx/* -1 (positive-diff y x)))))))
```

Comment:
I think 70 lines is just way too long for a Clojure function. It makes it hard to read. 

I suggested this in a different PR of yours last week but maybe you missed it -- why not just create a new multimethod in `sql.qp` to implement `datetime-diff` that dispatches off of driver + unit? The default impl of `->honeysql` would just hand off to the other method. Stuff like the recursive calls to `sql.qp/->honeysql` could be done in the generic implementation instead of being repeated each time for each driver.

Then instead of writing these mega methods everywhere you can just write a series of small method implementations for each unit. This is sort of like how the `sql.qp/date` method works. Even tho `sql.qp/date` isn't the best name in the world it's a lot easier to scan a bunch of small method implementations than it is to scan one mega one.

Reply:
I'm on the fence about making `allowed-database-diff-types` a method, btw. Writing some sort of helper function called  `check-allowed-datetime-diff-types` or something like that would be perfectly fine too

Reply:
Even for drivers where you can combine several different units into a single case statement I still think the separate method makes sense. Just abstract out the shared stuff into a helper function and write a bunch of one-line implementations for the units and hand off to it. That's what we do for the `sql.qp/date` function in a lot of places for example

Reply:
Thanks for the detailed reply, I've created a new PR [here](https://github.com/metabase/metabase/pull/27237) with this idea in mind. I'd appreciate a review from you before I refactor the rest of the drivers 🙏